src/lib/kcalid.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2015 Jolla Ltd. and/or its subsidiary(-ies).
+ * Contributors: Chris Adams <chris.adams@jollamobile.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include <QtDebug>
+#include <QString>
+#include <KDateTime>
+#include <incidence.h>
+
+#ifndef KCALID_H
+#define KCALID_H
+
+// Every incidence in the m/kcal database is uniquely identified by
+// a combination of the series UID and the occurrence's recurrenceId
+// which identifies it within that series.
+
+class KCalId
+{
+public:
+    KCalId() {} // default ctor needed for QHash semantics.
+    explicit KCalId(const QString &incidenceUid, const KDateTime &incidenceRecurrenceId = KDateTime())
+        : uid(incidenceUid), recurrenceId(incidenceRecurrenceId) {}
+    explicit KCalId(const KCalCore::Incidence::Ptr &incidence)
+        : uid(incidence->uid()), recurrenceId(incidence->recurrenceId()) {}
+
+    QString toString() const
+    {
+        // b64 encode the stringified recurrenceId to avoid colons.
+        return QStringLiteral("%1:%2")
+               .arg(uid)
+               .arg(recurrenceId.isValid() ? encodedRid(recurrenceId.toString()) : QString());
```

Comment:
That's why I use lastIndexOf() in the fromString() function, and base64 encode the recurrenceId string (because that ensures that no colon appears in it).  But yes, if there's some other character I could use without base64 encoding being required, that would be better.

More and more, I'm wondering about just adding a new column to the table, which contains the recurrence id, similarly to what I did for Google.  What do you think?